lib/ts/controllers/s-modal.ts

```diff
@@ -0,0 +1,279 @@
+            var triggeredEvent = this.triggerEvent(toShow ? "show" : "hide");
+
+            // if this pre-show/hide event was prevented, don't attempt to continue changing the modal state
+            if (triggeredEvent.defaultPrevented) {
+                return;
+            }
+
+            this.modalTarget.setAttribute("aria-hidden", toShow ? "false" : "true");
+
+            if (toShow) {
+                this.bindDocumentEvents();
+            }
+            else {
+                this.unbindDocumentEvents();
+                this.focusReturnElement();
+                this.removeModalOnHide();
+            }
+
+            // check for transitionend support
+            var supportsTransitionEnd = (<HTMLElement>this.modalTarget).ontransitionend !== undefined;
+
+            // shown/hidden events trigger after toggling the class
+            if (supportsTransitionEnd) {
+                // wait until after the modal finishes transitioning to fire the event
+                this.modalTarget.addEventListener("transitionend", () => {
+                    //TODO this is firing waaay to soon?
+                    this.triggerEvent(toShow ? "shown" : "hidden");
+                }, { once: true });
+            } else {
+                this.triggerEvent(toShow ? "shown" : "hidden");
+            }           
+        }
+
+        /**
+         * Listens for the s-modal:hidden event and focuses the returnElement when it is fired
+         */
+        private focusReturnElement() {
+            if (!this.returnElement) {
+                return;
+            }
+
+            this.modalTarget.addEventListener("s-modal:hidden", () => {
+                // double check the element still exists when the event is called
+                if (this.returnElement && document.body.contains(this.returnElement)) {
+                    this.returnElement.focus();
+                }
+            }, {once: true });
+        }
+
+        /**
+         * Remove the element on hide if the `remove-when-hidden` flag is set
+         */
+        private removeModalOnHide() {
+            if (this.data.get("remove-when-hidden") !== "true") {
+                return;
+            }
+
+            this.modalTarget.addEventListener("s-modal:hidden", () => {
+                this.element.remove();
+            }, {once: true });
+        }
+
+        /**
+         * Binds tab presses on tabbable items such that tabbing only works within the modal
+         */
+        private handleFocusableElements() {
+            // get all tabbable items
+            var allTabbables = Array.from(this.modalTarget.querySelectorAll("[href], input, select, textarea, button, [tabindex]"))
+                .filter((el: Element) => el.matches(":not([disabled]):not([tabindex='-1'])"));
+
+            if (!allTabbables.length) {
+                return;
+            }
+
+            var initialFocus = <HTMLElement>allTabbables[0];
+
+            if (this.hasInitialFocusTarget) {
```

Comment:
:+1: We have some use patterns where multiple elements may have `data-target="s-modal.initialFocus"` because we're looping or rendering conditionally.  I've double checked and this code won't blow up and will successfully match the first element in the list. https://codepen.io/bnickel/pen/abOVBmX?editors=1011